compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala

```diff
@@ -972,7 +972,7 @@ class ClassfileParser(
               def isStdlibClass(cls: ClassDenotation): Boolean =
                 ctx.platform.classPath.findClassFile(cls.fullName.mangledString) match {
                   case Some(entry: ZipArchive#Entry) =>
-                    entry.underlyingSource.map(_.name.startsWith("scala3-library_3-")).getOrElse(false)
+                    entry.underlyingSource.map(_.name.startsWith("scala3-library_")).getOrElse(false)
```

Comment:
I'm still not a fan of this kind of check, as I said jars can have any names and it'd be weird if we suddenly started imposing undocumented restriction on jar names in some situations.

Reply:
one situation where this could potentially fail: if someone uses sbt-assembly to package all their dependencies together and also uses staging, the compiler used at runtime will have only the sbt-assembly jar on its classpath which won't be called scala3-library

Reply:
@smarter Do you think of something like retrieving the information from the file `META-INF/MANIFEST.TXT` belonging to the *same jar file* as the class file ?

For instance, in `scala3-library_3-3.1.1-RC2.jar` we have : 
```
Manifest-Version: 1.0
Implementation-Title: scala3-library-bootstrapped
Automatic-Module-Name: org.scala.lang.scala3.library
Implementation-Version: 3.1.1-RC2
Specification-Vendor: LAMP/EPFL
Specification-Title: scala3-library-bootstrapped
Implementation-Vendor-Id: org.scala-lang
Specification-Version: 3.1.1-RC2
Implementation-URL: https://github.com/lampepfl/dotty
Implementation-Vendor: LAMP/EPFL
```

Reply:
Not sure if I understood you correctly. Let's consider a simple scenario.
Project A depends on project B.
Project B is packaged into sbt-assembly jar together with all of its dependencies (including stdlib)

Now we have a few possibilities of what version of the compiler A and B are compiled with:
a) Both A and B use the same minor version - OK
b) A uses 3.x1.y1 without -Yscala-release  and B uses 3.x2.y2 if x1 > x2 - OK no matter if B uses -Yscala-release
c) A uses 3.x1.y1 and B uses 3.x2.y2 with `-Yscala-release 3.x1` if x1 < x2 - here we encounter a problem because the fat jar will contain stdlib classes from 3.x2.y2 instead of 3.x1.y1  and 3.x1.y1 compiler won't be able to read them.
This isn't specifically a problem of this one check because the problem would still occur if x1 = 0 and x2 = 1. This seems all about the fact that build tools are not aware of -Yscala-release. But still we would need to release this feature as experimental first to let people play with it and then fix the build tools

Reply:
I don't have a specific scenario in mind, I just don't think it's wise to have a check like this.

Reply:
I really think we should just warn people to not use package scala, it's already dangerous because of package private scala definitions.

Reply:
(also the reason why we're special casing the standard library here should be documented in the code)

Reply:
I'm not opposed to having a filename-based check for now if it avoids accidentally using a 3.1 dependency from a -Yscala-release 3.0 project, but we should treat it as a temporary hack until we have a reliable way of detecting tasty files from the standard library (we could have a flag in the tasty file for that for example).

Reply:
Actually I thought about adding such kind of a flag to TASTy but for this proper solution we would have to wait until the next minor release. And even now in the worst case our checks would be too strict. So while checking TASTy version we make an exception if we're pretty sure some class comes from the stdlib while checking only the package prefix would allow some other libraries leak references to the newer stdlib API